src/actions/profile-view.js

```diff
@@ -269,154 +276,436 @@ export function changeSelectedThreads(
-            const counterSelectors = getCounterSelectors(counterIndex);
-            const counter = counterSelectors.getCounter(getState());
-            selectedThreadIndex = counter.mainThreadIndex;
-            break;
-          }
-          case 'event-delay': {
-            selectedThreadIndex = localTrack.threadIndex;
-            break;
-          }
-          case 'process-cpu':
-          case 'power': {
-            const { counterIndex } = localTrack;
-            const counterSelectors = getCounterSelectors(counterIndex);
-            const counter = counterSelectors.getCounter(getState());
-            selectedThreadIndex = counter.mainThreadIndex;
-            break;
-          }
-          default:
-            throw assertExhaustiveCheck(
-              localTrack,
-              `Unhandled LocalTrack type.`
-            );
-        }
-        break;
+        dispatch({
+          type: 'SELECT_TRACK',
+          selectedThreadIndexes: newSelectedThreadIndexes,
+          selectedTab,
+          // In this case, we keep the old information.
+          // This allows the user to do shift+click again to cancel the last
+          // shift+click and do it again.
+          lastNonShiftClickInformation,
+        });
+        return;
       }
-      default:
-        throw assertExhaustiveCheck(
-          trackReference,
-          'Unhandled TrackReference type'
-        );
     }
 
-    const visibleTabs = getThreadSelectors(selectedThreadIndex).getUsefulTabs(
-      getState()
+    // If we couldn't select a range for some reason, then select just the
+    // clicked track. Most likely the user just loaded a profile with several
+    // selected tracks, and tried to use shift-click immediately.
+    dispatch(setOneTrackSelection(clickedTrackInformation, selectedTab));
+  };
+}
+
+/**
+ * This selects a track from its reference.
+ * This will ultimately select the thread that this track belongs to, using its
+ * thread index, and may also change the selected tab if it makes sense for this
+ * track.
+ */
+export function selectTrackWithModifiers(
+  trackReference: TrackReference,
+  modifiers: $Shape<{ ctrlOrMeta: boolean, shift: boolean }> = {}
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    // These get assigned based on the track type.
+    const clickedTrackInformation = getInformationFromTrackReference(
+      getState(),
+      trackReference
     );
+    if (!clickedTrackInformation) {
+      // This track isn't selectable.
+      return;
+    }
+
+    let selectedTab =
+      clickedTrackInformation.relatedTab ?? getSelectedTab(getState());
+    const visibleTabs = getThreadSelectors(
+      clickedTrackInformation.relatedThreadIndex
+    ).getUsefulTabs(getState());
     if (!visibleTabs.includes(selectedTab)) {
       // If the user switches to another track that doesn't have the current
       // selectedTab then switch to the first tab.
       selectedTab = visibleTabs[0];
     }
```

Comment:
I wasn't sure what to do with the `selectedTab`, should we change it for all modes? Maybe we'd want to change it only for `setOneTrackSelection`... I thought it would be easy to change in a follow-up if needed.